File: 543. Diameter of Binary Tree/543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int res = -1;
    int dfs(TreeNode* root){
        if(!root) return 0; // if null, end traversal.
        int L = dfs(root->left);
        int R = dfs(root->right);
        res = std::max(L + R, res); // maintain the max path while traversal.
        return std::max(L, Rc) + 1; // return the depth of this tree.
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        dfs(root);
        return res;
    }
};

File: 543. Diameter of Binary Tree/543. Diameter of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res = -1;
    private int dfs(TreeNode root){
        if(root == null) return 0; // if null, end traversal.
        int L = dfs(root.left);
        int R = dfs(root.right);
        res = Math.max(L + R, res); // maintain the max path while traversal.
        return Math.max(L, R) + 1; // return the depth of this tree.
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null) return 0;
        dfs(root);
        return res;
    }
}